Add new tiltrotor flightmodes

kd0aij opened the pull request:
QACRO and QSTABILIZE with mixing from throttle to forward throttle 

Todo: rear motor output never drops below 1500usec; maybe shut it down at higher speeds

Hwurzburg commented:
flight tested today in QSFT mode...windy....tried all values of THR_MAX...used default mix of 0.5...I actually liked 1.0 the best...no problem transitioning to/from CRUISE mode (should be the same for all plane modes)
however, I noticed that the motors tilted forward before 50% throttle....this causes it to start scooting along the ground forward before hover....didnt like that at all...I think it should be exactly like QSTAB up to 50% throttle, then engage forward thrust...

kd0aij commented:
Hi Henry,
I noticed that the motor tilt on my big Convergence was different from the
mini (and same as you describe) after I got out to the field today, but
wasn't able to see what was different between the two.
I had tested yesterday on the Mini, which is an mRoX2.1 (fmuv3), and the
motors don't start tilting till the throttle stick is above 50%. Figuring
out why the behavior is different is at the top of my todo list...

But I didn't like the way QSFT felt on the mini; I either had too much or
too little throttle to maintain altitude at low speeds. That wasn't a
problem when using the extra RC channel to directly control the forward
tilt.

Do you think the new modes with mixing are worth pursuing, based on your
flights today?

thanks,
--Mark

On Tue, Nov 19, 2019 at 12:56 PM Henry Wurzburg <notifications@github.com>
wrote:

> flight tested today in QSFT mode...windy....tried all values of
> THR_MAX...used default mix of 0.5...I actually liked 1.0 the best...no
> problem transitioning to/from CRUISE mode (should be the same for all plane
> modes)
> however, I noticed that the motors tilted forward before 50%
> throttle....this causes it to start scooting along the ground forward
> before hover....didnt like that at all...I think it should be exactly like
> QSTAB up to 50% throttle, then engage forward thrust...
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ArduPilot/ardupilot/pull/12857?email_source=notifications&email_token=AARRSPPLV72V2UFBCTMS3W3QURAHXA5CNFSM4JO4LQS2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEPQ4GY#issuecomment-555683355>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARRSPLTRG4G6ILWNCTUFQLQURAHXANCNFSM4JO4LQSQ>
> .
>


-- 
Mark Whitehorn
kd0aij@gmail.com

Hwurzburg commented:
Absolutely...I would probably use QSFT instead of QSTAB if we can get the tilt to stay put until over hover throttle....I would be happy to use mix=.5 and default the thr_max for the  no rcx_option=78 case to be 1.0, full max throttle....I do think that lowering rear motor toward zero as you approach full throttle would be more efficient, but I didnt notice a bad flight effect as is....roll limit needs to be addressed, its a pig in turns at forward flight speeds...I always increase FBWA roll limits to 50-60deg for wings...need to be able to do the same in this mode somehow...either use the plane limits or add one for this mode

kd0aij commented:
For some reason, it seems roll and pitch are limited by:
LIM_PITCH_MAX    6000.000000
LIM_PITCH_MIN    -6000.000000
LIM_ROLL_CD      6000.000000

Those are the settings I used in SITL.

Hwurzburg commented:
Are you sure it isn't q_tilt_max since it's in vtol state?

kd0aij commented:
It's coming from here: https://github.com/ArduPilot/ardupilot/blob/master/ArduPlane/mode_qstabilize.cpp#L21

Hwurzburg commented:
Using the lowest one..plane or vtol....so I think it needs a separate one just for these modes....like acro has a separate rate adjustment

kd0aij commented:
This graph is from SITL, showing that tilt (supposedly) begins at hover throttle:
![image](https://user-images.githubusercontent.com/2300221/69262281-70897f00-0b80-11ea-9f86-35283be2e85a.png)

kd0aij commented:
@Hwurzburg I found the issue with the tilt mixing. It was dependent on throttle trim.
I fixed that and changed the function of the RC fwd throttle channel to offset the start of tilt by +/-0.25
(if assigned). I also changed the roll limit to MAX instead of MIN; maybe we can get by without new parameter #1350 :)

Hwurzburg commented:
I can see that without RC fwd channel enabled, it behaves well with nothing beginning until hover and mix changing max fwd from 0 to 1 directly, at full throttle stick....but when the RC channel is used it seems to change not only the max fwd throttle limit, but when the tilt begins...so  anytime the slider is off zero, will have the vehicle scooting forward on the ground before liftoff.....at full slider tilt starts at 25% throttle....it does increase the max fwd throttle: with .5 mix it increases it from 50% to 75% at full stick ...and with 1.0 mix it reaches fwd throttle=1 at 3/4 stick throttle...seems like it might be a bit confusing? with RC fwd channel enabled, I would expect it to control thr_max, with mix controlling the curve of fwd throttle vs stick above 50% ,but still not start until 50% stick

kd0aij commented:
I figured the (now misnamed) rc_fwd_thr slider would be useful for tweaking the start of tilt around mid-stick to see whether we want an extra parameter to set an offset. I didn't try to compensate the max tilt as a function of offset, since this is just for testing.

kd0aij commented:
@Hwurzburg I tried maneuvering close to the ground in QSFTHR again on a gusty day, and couldn't find a position I liked for the slider; depending on the wind and vehicle heading, I kept needing to adjust throttle, pitch and the slider to hold position and altitude.
I find it easier to use the slider to directly control tilt (forward throttle) without the mixing of throttle into tilt. Perhaps it would be better to try this with an altitude controlled mode instead. I can't think of a way to improve the mix from throttle to tilt.

peterbarker commented:
We've merged something which gives equivalent functionality to this, `Q_FWD_THR_USE`.

Thanks anyway!
